### PR TITLE
Keep QDQ nodes w/ nonpositive scale around MaxPool

### PR DESCRIPTION
**Pull request by mcollinswisc**
### Description
This change adds a check for whether the scale in the QuantizeLinear (or DequantizeLinear) is a positive scalar, and a new selector to disallow removing the QDQ around MaxPool if it is not.

### Motivation and Context
Currently, the DropQDQNodesRules optimization removes QuantizeLinear and DequantizeLinear nodes from DequantizeLinear ∘ MaxPool ∘ QuantizeLinear. However, if the x_scale/y_scale values are non-positive, the (de-)quantization changes the ordering of the elements in the input value, so this optimization is changing the results.

https://github.com/microsoft/onnxruntime/issues/21176

### CODE REVIEW

**Comment by mcollinswisc** [comment]
@microsoft-github-policy-service agree company="TetraMem"

File: onnxruntime/core/optimizer/qdq_transformer/selectors_actions/qdq_selector_action_transformer.cc [hunk 1]
```diff
@@ -46,19 +47,31 @@ void DropQDQNodesRules(SelectorActionRegistry& qdq_selector_action_registry) {
...
-                                                          {"Resize", {}}},
+                                                         {{"Resize", {}}},
                                                          std::move(selector_disallow_16bit),
                                                          std::move(drop_action_no_int16));
 
+  std::unique_ptr<NodeSelector> selector_disallow_16bit_and_nonpositive_scale = std::make_unique<QDQ::DropQDQNodesSelector>(false, true, false);
+  qdq_selector_action_registry.RegisterSelectorAndAction(drop_action_no_int16_nor_nonpositive_scale_name,
+                                                         {{"MaxPool", {12}}},
+                                                         std::move(selector_disallow_16bit_and_nonpositive_scale),
+                                                         std::move(drop_action_no_int16_nor_nonpositive_scale));
```
Comment (edgchen1) [inline onnxruntime/core/optimizer/qdq_transformer/selectors_actions/qdq_selector_action_transformer.cc:73]: this new action also needs to be registered with the same name using RegisterAction() in the minimal build code path (other part of the ifdef for `defined(ORT_MINIMAL_BUILD)`).

File: onnxruntime/core/optimizer/qdq_transformer/selectors_actions/qdq_selector_action_transformer.cc [hunk 2]
```diff
@@ -46,19 +47,31 @@ void DropQDQNodesRules(SelectorActionRegistry& qdq_selector_action_registry) {
 
   std::unique_ptr<Action> drop_action_no_int16 = std::make_unique<MergeIntoTargetFixed>(
       std::vector<NodeAndMoveInfo>(moves));  // Copy before std::move(moves)
+  std::unique_ptr<Action> drop_action_no_int16_nor_nonpositive_scale = std::make_unique<MergeIntoTargetFixed>(
+      std::vector<NodeAndMoveInfo>(moves));  // Copy before std::move(moves)
   std::unique_ptr<Action> drop_action = std::make_unique<MergeIntoTargetFixed>(std::move(moves));
 
 #if !defined(ORT_MINIMAL_BUILD)
-  // Use a separate selector + action that disallows 16-bit types for MaxPool and Resize.
+  // Use a separate selectors & actions for MaxPool and Resize.
```
Comment (edgchen1) [inline onnxruntime/core/optimizer/qdq_transformer/selectors_actions/qdq_selector_action_transformer.cc:55]:
```suggestion
  // Use separate selectors & actions for MaxPool and Resize.
```

**Comment by mcollinswisc** [comment]
Also incorporating some name changes from scottmckay on stacked PR https://github.com/microsoft/onnxruntime/pull/21376#discussion_r1690768042 (thanks to both for review)

File: onnxruntime/core/optimizer/qdq_transformer/selectors_actions/qdq_selector_action_transformer.cc (hunk 1 above)
Comment (mcollinswisc) [reply onnxruntime/core/optimizer/qdq_transformer/selectors_actions/qdq_selector_action_transformer.cc]: I believe this is done in
https://github.com/microsoft/onnxruntime/pull/21182/commits/023bbafa10e429af208a2c613d869ef9ee335085

File: onnxruntime/test/optimizer/qdq_transformer_test.cc [hunk 3]
```diff
@@ -787,7 +787,7 @@ void QDQTransformerGemmTests(bool has_output_q, bool has_bias, bool beta_not_one
     auto check_binary_op_graph = [&](InferenceSessionWrapper& session) {
       auto op_to_count = CountOpsInGraph(session.GetGraph());
       const QDQOpKeys qdq_keys = GetQDQOpKeys(use_contrib_qdq);
-      if ((!has_output_q || std::is_same_v<Input1Type, OutputType>) && (!has_bias || (std::is_same_v<BiasType, int32_t> && !beta_not_one)) &&
+      if ((!has_output_q || std::is_same_v<Input1Type, OutputType>)&&(!has_bias || (std::is_same_v<BiasType, int32_t> && !beta_not_one)) &&
```
Comment (edgchen1) [inline onnxruntime/test/optimizer/qdq_transformer_test.cc:790]: the formatting check will complain about this. you can run the formatter locally - clang-format or lintrunner.
https://github.com/microsoft/onnxruntime/blob/main/docs/Coding_Conventions_and_Standards.md#linting

**Review by edgchen1 — APPROVED** [review]

**Pull request merged**